neoemf-core/src/main/java/fr/inria/atlanmod/neoemf/data/BackendFactoryRegistry.java

```diff
@@ -0,0 +1,204 @@
+import fr.inria.atlanmod.neoemf.bind.Bindings;
+import fr.inria.atlanmod.neoemf.bind.FactoryBinding;
+import fr.inria.atlanmod.neoemf.data.store.Store;
+import fr.inria.atlanmod.neoemf.resource.PersistentResource;
+import fr.inria.atlanmod.neoemf.resource.PersistentResourceFactory;
+import fr.inria.atlanmod.neoemf.util.UriBuilder;
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.resource.Resource;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import static fr.inria.atlanmod.commons.Preconditions.checkNotNull;
+import static java.util.Objects.isNull;
+import static java.util.Objects.nonNull;
+
+/**
+ * The registry that holds registered {@link URI} schemes with their associated {@link BackendFactory}.
+ * <p>
+ * This {@code BackendFactoryRegistry} is used for dynamically create {@link Store} and {@link Backend} when loading and
+ * saving a {@link PersistentResource}. For this reason, a {@link BackendFactory} must be registered before using these
+ * operations, with the {@link #register(String, BackendFactory)} method.
+ *
+ * @see PersistentResource#load(Map)
+ * @see PersistentResource#save(Map)
+ */
+@Singleton
+@ParametersAreNonnullByDefault
+public final class BackendFactoryRegistry {
+
+    /**
+     * A map containing all registered {@link BackendFactory} identified by a {@link URI} scheme.
+     */
+    @Nonnull
+    private final Map<String, BackendFactory> factories = new ConcurrentHashMap<>();
+
+    /**
+     * Whether this registry has been initialized at least once.
+     */
+    private boolean initialized;
+
+    /**
+     * Constructs a new {@code BackendFactoryRegistry}.
+     */
+    private BackendFactoryRegistry() {
+    }
+
+    /**
+     * Returns the instance of this class.
+     *
+     * @return the instance of this class
+     */
+    @Nonnull
+    public static BackendFactoryRegistry getInstance() {
+        return Holder.INSTANCE;
+    }
+
+    /**
+     * Returns all registered {@link URI} schemes with their {@link BackendFactory}.
+     *
+     * @return an immutable map
+     */
+    @Nonnull
+    @VisibleForTesting
+    public Map<String, BackendFactory> getFactories() {
+        if (!initialized) {
+            registerAll();
+        }
+
+        return Collections.unmodifiableMap(factories);
+    }
+
+    /**
+     * Returns the {@link BackendFactory} identified by the given {@link URI} {@code scheme}.
+     *
+     * @param scheme the {@link URI} scheme identifying the factory
+     *
+     * @return the factory
+     *
+     * @throws NullPointerException if {@code scheme} is {@code null}, or if no factory is registered for the {@code
+     *                              scheme}
+     */
+    @Nonnull
+    public BackendFactory getFactoryFor(String scheme) {
+        checkNotNull(scheme, "scheme");
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "scheme" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)